Introduce guidelines to setup Supabase with Android Kotlin including sample with Jetpack Compose (#14571)
supabase/supabase: apps/docs/pages/guides/getting-started/tutorials/with-kotlin.mdx

```kotlin
val url = BuildConfig.SUPABASE_URL
val apiKey = BuildConfig.API_KEY
```

### 3. Set up Supabase dependencies including Postgrest, GoTrue, Storage and serialization plugin
<img width="720" alt="Screen Shot 2023-06-05 at 21 39 24" src="https://github.com/hieuwu/product-sample-supabase-kt/assets/43868345/3f154fda-aa9e-4baa-81f2-5b744c7ad2fe"></img>

In build.gradle (app) file, add these dependencies then press Sync now. Replace the dependency version placeholders `$supabase_version` and `$ktor_version` with the latest versions as shown below.
    <div style={{ display: 'flex', flexDirection: 'row', alignItems: 'center' }}> 
      <a href="https://github.com/supabase-community/supabase-kt/releases" style={{ marginRight: '8px' }}> 
          <img src="https://img.shields.io/github/release/supabase-community/supabase-kt?label=supabase-kt stable" alt="latest stable supabase-kt version"/> 
      </a> 
      <a href="https://github.com/supabase-community/supabase-kt/releases"> 
          <img src="https://img.shields.io/github/v/release/ktorio/ktor.svg?label=ktor" alt="latest ktor version"/> 
      </a> 
    </div> 
```kotlin
// Supabase setup

> implementation "io.github.jan-tennert.supabase:postgrest-kt-android:$supabase_version"

Reviewer: GoTrue dependency missing

Author: Updated

                                        navController.navigate(
                                            ProductDetailsDestination.createRouteWithParam(
                                                item.id
                                            )
                                        )
                                    },
                                )
                            },
                            directions = setOf(DismissDirection.EndToStart),
                        )
                    }
                }
            } else {
                Text("Product list is empty!")
            }

        }
    }
}

@Composable
private fun AddProductButton(
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
) {
    FloatingActionButton(
        modifier = modifier,
        onClick = onClick,
        containerColor = MaterialTheme.colorScheme.primary,
        contentColor = MaterialTheme.colorScheme.onPrimary
    ) {
        Icon(
            imageVector = Icons.Filled.Add,
            contentDescription = null,
        )
    }
}
```

Product details
Create ProductDetailsViewModel.kt
```kotlin

@HiltViewModel

> class ProductDetailsViewModel @Inject constructor(
>     private val productRepository: ProductRepository,
>     savedStateHandle: SavedStateHandle,
> 
>     ) : ViewModel() {

Reviewer: ```suggestion
class ProductDetailsViewModel @Inject constructor(
    private val productRepository: ProductRepository,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {
```

                            duration = SnackbarDuration.Long
                        )
                    }
                }) {
                Text("Sign up")
            }
        }
    }
}
```
Sign In screen
Create SignInViewModel
```kotlin
@HiltViewModel
class SignInViewModel @Inject constructor(
    private val authenticationRepository: AuthenticationRepository
) : ViewModel() {

    private val _email = MutableStateFlow("")
    val email: Flow<String> = _email

    private val _password = MutableStateFlow("")
    val password = _password

    fun onEmailChange(email: String) {
        _email.value = email
    }

    fun onPasswordChange(password: String) {
        _password.value = password
    }

    fun onLogin() {
        viewModelScope.launch {
            authenticationRepository.signIn(
                email = _email.value,
                password = _password.value
            )
        }
    }
}
```

Create SignInScreen.kt
```kotlin
@OptIn(ExperimentalMaterial3Api::class, ExperimentalComposeUiApi::class)
@Composable
fun SignInScreen(
    modifier: Modifier = Modifier,
    navController: NavController,

>     viewModel: SignInViewModel = hiltViewModel()
> ) {
> 
>     val snackBarHostState = remember { SnackbarHostState() }

Reviewer: ```suggestion
    viewModel: SignInViewModel = hiltViewModel()
) {
    val snackBarHostState = remember { SnackbarHostState() }
```

}
```
### 5. Provide Supabase instances with Hilt
You can create it by just calling the constructor. However to make the app more testable, these should be provided
Create SupabaseMdule.kt
```kotlin
@InstallIn(SingletonComponent::class)
@Module
object SupabaseModule {

    @Provides
    @Singleton
    fun provideSupabaseClient(): SupabaseClient {
        return createSupabaseClient(
            supabaseUrl = BuildConfig.SUPABASE_URL,
            supabaseKey = BuildConfig.API_KEY
        ) {
            install(Postgrest)
            install(GoTrue)
            install(Storage)
        }
    }

    @Provides
    @Singleton
    fun provideSupabaseDatabase(client: SupabaseClient): Postgrest {
        return client.postgrest
    }
    
    @Provides
    @Singleton
    fun provideSupabaseGoTrue(client: SupabaseClient): GoTrue {
        return client.gotrue
    }

    @Provides
    @Singleton
    fun provideSupabaseStorage(client: SupabaseClient): Storage {
        return client.storage
    }

}
```

### 6. Create data transfer object (DTO) to get data
Create `ProductDto.kt` class and use annotations to parse data from Supabase
```kotlin
@Serializable

> data class ProductDto(

Reviewer: I suppose this part is not completely done, but in the preview, there is sql code to create a table for user profiles and not products. Also, is there a reason that `image` and `id` are nullable, but default to an empty string?

Author: I misused the sql query from other sample, I updated with mine. My bad for making the id nullable just updated.
My purpose of making `image` nullable is that at the beginning, we might not create a product with an image. After that, we can update that image.

Reviewer: All good, it was just irritating that they both default to an empty string rather than null as they were nullable

Author: You are correct. It should be null instead of empty string. Updated for all places

    <div style={{ display: 'flex', flexDirection: 'row', alignItems: 'center' }}> 
      <a href="https://github.com/supabase-community/supabase-kt/releases" style={{ marginRight: '8px' }}> 
          <img src="https://img.shields.io/github/release/supabase-community/supabase-kt?label=supabase-kt stable" alt="latest stable supabase-kt version"/> 
      </a> 
      <a href="https://github.com/supabase-community/supabase-kt/releases"> 
          <img src="https://img.shields.io/github/v/release/ktorio/ktor.svg?label=ktor" alt="latest ktor version"/> 
      </a> 
    </div> 
```kotlin
// Supabase setup
implementation "io.github.jan-tennert.supabase:postgrest-kt-android:$supabase_version"
implementation "io.github.jan-tennert.supabase:storage-kt:$supabase_version"
implementation "io.ktor:ktor-client-android:$ktor_version"
implementation "io.ktor:ktor-client-logging-jvm:$ktor_version"
implementation "io.ktor:ktor-client-core:$ktor_version"
implementation "io.ktor:ktor-client-cio:$ktor_version"
implementation "io.ktor:ktor-utils-jvm:$ktor_version"
```

Also in build.gradle (app) file, add plugin for serialization
```kotlin
plugins {
    ...

>     id 'org.jetbrains.kotlin.plugin.serialization' version '1.7.0'

Reviewer: Same as above

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  title: 'Build a Product Management App with Jetpack Compose in Android',
  description: 'Learn how to use Supabase in your Android Kotlin App.',
}

> <QuickstartIntro />

Reviewer 2: The quickstart intro mentions Supabase Auth + Storage. However this guide only focus on Supabase db.

We should implement login flow + profile photo update to showcase Auth + Storage. wdyt @hieuwu

Author: Actually I plan to create more screens in a separated tab with more action to make use of most APIs interacting with Auth and Storage (resuming, error handling,...). In this first iteration, I focus mostly on the database cause this requires lots of steps and setup. I will introduce Auth + Storage in next iterations. Is this approach okay @phamhieu ?

Author: You are right. I will introduce these. It won't take much time actually

![manage-product-cover](https://github.com/hieuwu/product-sample-supabase-kt/assets/43868345/1bed0c84-208a-4266-a2ec-2bac50ddf80c)

<Admonition type="note">
  If you get stuck while working through this guide, refer to the [full example on

>   GitHub](https://github.com/hieuwu/product-sample-supabase-kt).

Reviewer 2: copy this to https://github.com/supabase/supabase/tree/master/examples for consistent

```kotlin
defaultConfig {
   applicationId "com.hieuwu.groceriesstore"
   minSdkVersion 22
   targetSdkVersion 33
   versionCode 5
   versionName "1.0"
   testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"

   // Set value part
   Properties properties = new Properties()
   properties.load(project.rootProject.file("local.properties").newDataInputStream())
   buildConfigField("String", "API_KEY", "\"${properties.getProperty("API_KEY")}\"")
   buildConfigField("String", "SECRET", "\"${properties.getProperty("SECRET")}\"")
   buildConfigField("String", "SUPABASE_URL", "\"${properties.getProperty("SUPABASE_URL")}\"")
}
```

#### 2.3. Use value from `BuildConfig`
Read value from `BuildConfig` (Remember to import the right class, it is one from your project). For example:
```kotlin
val url = BuildConfig.SUPABASE_URL
val apiKey = BuildConfig.API_KEY
```

### 3. Set up Supabase dependencies including Postgrest, GoTrue, Storage and serialization plugin
<img width="720" alt="Screen Shot 2023-06-05 at 21 39 24" src="https://github.com/hieuwu/product-sample-supabase-kt/assets/43868345/3f154fda-aa9e-4baa-81f2-5b744c7ad2fe"></img>

> In build.gradle (app) file, add these dependencies then press Sync now. Replace place holder of the dependency versions `$supabase_version` and `$ktor_version` with latest versions as below

Reviewer: ```suggestion
In build.gradle (app) file, add these dependencies then press Sync now. Replace the dependency version placeholders `$supabase_version` and `$ktor_version` with the latest versions as shown below.
```

      <a href="https://github.com/supabase-community/supabase-kt/releases" style={{ marginRight: '8px' }}> 
          <img src="https://img.shields.io/github/release/supabase-community/supabase-kt?label=supabase-kt stable" alt="latest stable supabase-kt version"/> 
      </a> 
      <a href="https://github.com/supabase-community/supabase-kt/releases"> 
          <img src="https://img.shields.io/github/v/release/ktorio/ktor.svg?label=ktor" alt="latest ktor version"/> 
      </a> 
    </div> 
```kotlin
// Supabase setup
implementation "io.github.jan-tennert.supabase:postgrest-kt-android:$supabase_version"
implementation "io.github.jan-tennert.supabase:storage-kt:$supabase_version"
implementation "io.ktor:ktor-client-android:$ktor_version"
implementation "io.ktor:ktor-client-logging-jvm:$ktor_version"
implementation "io.ktor:ktor-client-core:$ktor_version"
implementation "io.ktor:ktor-client-cio:$ktor_version"
implementation "io.ktor:ktor-utils-jvm:$ktor_version"
```

Also in build.gradle (app) file, add plugin for serialization. The version of this plugin should be the same with Kotlin version
<a href="https://github.com/JetBrains/kotlin" style={{ marginRight: '8px' }}> 

>         <img src="https://img.shields.io/maven-central/v/org.jetbrains.kotlin/kotlin-maven-plugin.svg?label=kotlin_version" alt="latest stable supabase-kt version"/> 

Reviewer: ```suggestion
        <img src="https://img.shields.io/maven-central/v/org.jetbrains.kotlin/kotlin-maven-plugin.svg?label=kotlin_version" alt="latest kotlin version"/> 
```

                    set("name", name)
                    set("price", price)
                }) {
                    eq("id", id)
                }
            }
        }
    }

    // Because I named the bucket as "Product Image" so when it turns to an url, it is "%20"
    // For better approach, you should create your bucket name without space symbol
    private fun buildImageUrl(imageFileName: String) =
        "${BuildConfig.SUPABASE_URL}/storage/v1/object/public/${imageFileName}".replace(" ", "%20")
}
```

### 8. Implement screens
<Admonition type="note">
  By the time this page is written, there might be some changes in the code to make it more simple. If you face any problem when implement, feel free to refer the example repository above
</Admonition>

#### 8.1 Product list
Create ProductListViewModel
```kotlin
@HiltViewModel
class ProductListViewModel @Inject constructor(
    private val productRepository: ProductRepository,
) : ViewModel() {

    private val _productList = MutableStateFlow<List<Product>?>(listOf())
    val productList: Flow<List<Product>?> = _productList

    private val _isLoading = MutableStateFlow(false)

>     val isLoading: Flow<Boolean> = _isLoading

Reviewer 2: Can we only show part of the code block with collapse and expand button? 

cc @gregnr

Reviewer 3: @phamhieu Unfortunately no built-in expand functionality atm (but good idea!). Might be able to achieve something similar with one of Code Hike's components like [Spotlight](https://codehike.org/demo/spotlight). I would probably opt to leave this as is for now until we have a good pattern for this.
